plan/plan.go

```diff
@@ -52,36 +52,47 @@ type Changes struct {
 // processing. It returns a copy of Plan with the changes populated.
 func (p *Plan) Calculate() *Plan {
 	changes := &Changes{}
-
-	// Ensure all desired records exist. For each desired record make sure it's
-	// either created or updated.
-	for _, desired := range p.Desired {
-		// Get the matching current record if it exists.
-		current, exists := recordExists(desired, p.Current)
-
-		// If there's no current record create desired record.
-		if !exists {
-			changes.Create = append(changes.Create, desired)
-			continue
+	usedDesired := map[int]bool{}
+	usedCurrent := map[int]bool{}
+
+	for c, current := range p.Current {
+		for d, desired := range p.Desired {
+			if current.DNSName == desired.DNSName && current.Target == desired.Target &&
+				current.SuitableType() == desired.SuitableType() {
+				usedDesired[d] = true
+				usedCurrent[c] = true
+				log.Debugf("Skipping endpoint %v because target has not changed", desired)
+			}
 		}
+	}
 
-		// If there already is a record update it if it changed.
-		if desired.Target != current.Target {
-			changes.UpdateOld = append(changes.UpdateOld, current)
-
-			desired.RecordType = current.RecordType // inherit the type from the dns provider
-			desired.MergeLabels(current.Labels)     // inherit the labels from the dns provider, including Owner ID
-			changes.UpdateNew = append(changes.UpdateNew, desired)
+	for c, current := range p.Current {
+		if _, found := usedCurrent[c]; found {
 			continue
 		}
-
-		log.Debugf("Skipping endpoint %v because target has not changed", desired)
+		for d, desired := range p.Desired {
```

Comment:
yes, what you mentioned is already known. If we merge this fix then current external-dns will stop functioning, if it already has clashing CNAMEs, while the scenario u are mentioning can only happen if two CNAMEs were created within a time-frame of 1min (sync period, or if external dns was first run when two such resources were already created). So I propose before merging this to have a safety mechanism ensuring that batch requests do not fail in case if plan produced entities are not valid (e.g. two CNAMEs with same name). I am open for proposal regarding that